_posts/2024-06-14-granite-rag.adoc

```diff
@@ -0,0 +1,286 @@
+To run Granite locally, we'll be utilizing Instructlab, though Podman AI Studio is another viable option.
+
+The core of our application is based on the RAG (Retrieval-Augmented Generation) pattern.
+This approach enhances the chatbot's responses by retrieving pertinent information from a vector database — in this case, Infinispan — before generating a response.
+
+
+== Architecture
+
+Our chatbot application is composed of four main components:
+
+1. **Web Socket Endpoint**: This component serves as the communication bridge between the chatbot's backend and the frontend interface.
+This component uses the new Quarkus WebSocket-Next extension to handle WebSocket connections efficiently.
+It relies on the AI Service to interact with the LLM.
+
+2. **Ingestor**: The ingestor is responsible for populating the database with relevant data. It processes a set of local document, split them into text segment, compute their vector representation and store them into Infinispan.
+
+3. **Retriever**: The retriever allows finding relevant text segment into Infinispan. When a user inputs a query, the retriever searches the vector database to find the most relevant pieces of information.
+
+4. **AI Service**: This is the core component of the chatbot, combining the capabilities of the retriever and the Granite LLM. The AI service takes the relevant information fetched by the retriever and uses the Granite LLM to generate the appropriate responses.
+
+The following picture illustrates the high-level architecture:
+
+image::chatbot-architecture.png[Architecture of the application,float="right",align="center"]
+
+== A simple explanation of the RAG pattern
+
+The RAG (Retrieval-Augmented Generation) pattern is one of the most popular AI patterns, combining a retrieval mechanism with a generation mechanism to provide more accurate and relevant responses.
+
+The RAG pattern operates in two main steps:
+
+1. **Ingestion**: The application ingests a set of documents, processes them, and stores them in a vector database.
+2. **Retrieval**: When a user inputs a query, the application retrieves the most relevant information from the vector database.
+
+The following image gives a high-level overview of the _traditional_ RAG pattern:
+
+image::traditional-rag-pattern.png[The traditional RAG pattern,float="right",align="center"]
+
+There are more advanced version of the RAG pattern, but let's stick to the basics for this application.
+
+=== Ingestion
+
+Let's first look at the ingestion step.
+The ingestion process involves reading a set of documents, splitting them into text segments, computing their vector representations, and storing them in Infinispan.
+
+The secret of an effective RAG implementation lies in how the text segments are computed.
+In our application, we will follow a straightforward approach, but more advanced techniques are available and often required.
```

Comment:
Do you have pointers to these more advanced techniques? Might be interesting to add links.

Reply:
Added in the text.